syne_tune/optimizer/schedulers/searchers/kde_searcher.py

```diff
@@ -0,0 +1,275 @@
+        for hname in self.configspace:
+            hp = self.configspace[hname]
+            if isinstance(hp, sp.Categorical):
+                self.vartypes.append(('u', len(hp.categories)))
+            if isinstance(hp, sp.Integer):
+                self.vartypes.append(('o', (hp.lower, hp.upper)))
+            if isinstance(hp, sp.Float):
+                self.vartypes.append(('c', 0))
+
+        self.num_min_data_points = len(self.vartypes) if num_min_data_points is None else num_min_data_points
+        assert self.num_min_data_points >= len(self.vartypes)
+
+    @staticmethod
+    def to_feature(configspace, config, categorical_maps):
+        def numerize(value, domain, categorical_map):
+            if isinstance(domain, sp.Categorical):
+                res = categorical_map[value] / len(domain)
+                return res
+            else:
+                return [(value - domain.lower) / (domain.upper - domain.lower)]
+
+        return np.hstack([
+            numerize(value=config[k], domain=v, categorical_map=categorical_maps.get(k, {}))
+            for k, v in configspace.items()
+            if isinstance(v, sp.Domain)
+        ])
+
+    @staticmethod
+    def from_feature(configspace, feature_vector, inv_categorical_maps):
+        def inv_numerize(values, domain, categorical_map):
+            if not isinstance(domain, sp.Domain):
+                # constant value
+                return domain
+            else:
+                if isinstance(domain, sp.Categorical):
+                    index = int(values * len(domain))
+                    return categorical_map[index]
+                else:
+                    if isinstance(domain, sp.Float):
+                        return values * (domain.upper - domain.lower) + domain.lower
+                    else:
+                        return int(values * (domain.upper - domain.lower) + domain.lower)
+
+        res = {}
+        curr_pos = 0
+        for k, domain in configspace.items():
+            if hasattr(domain, "sample"):
+                res[k] = domain.cast(
+                    inv_numerize(
+                        values=feature_vector[curr_pos],
+                        domain=domain,
+                        categorical_map=inv_categorical_maps.get(k, {})
+                    )
+                )
+                curr_pos += 1
+            else:
+                res[k] = domain
+        return res
+
+    def to_objective(self, result: Dict):
+        if self.mode == 'min':
+            return result[self.metric_name]
+        elif self.mode == 'max':
+            return -result[self.metric_name]
+
+    def _update(self, trial_id: str, config: Dict, result: Dict):
+        self.X.append(self.to_feature(
+            config=config,
+            configspace=self.configspace,
+            categorical_maps=self.categorical_maps,
+        ))
+        self.y.append(self.to_objective(result))
+        self.train_kde(np.array(self.X), np.array(self.y))
```

Comment:
You should only do this in get_config, when you really need it.

Reply:
you mean training the KDE? I agree